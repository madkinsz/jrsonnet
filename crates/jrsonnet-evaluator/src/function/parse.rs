use jrsonnet_gcmodule::Trace;
use jrsonnet_interner::IStr;
use jrsonnet_parser::{LocExpr, ParamsDesc};

use super::{
	arglike::ArgsLike,
	builtin::{BuiltinParam, BuiltinParamName},
};
use crate::{
	destructure::destruct,
	error::{Error::*, Result},
	evaluate_named,
	gc::GcHashMap,
	tb, throw,
	val::ThunkValue,
	Context, Pending, State, Thunk, Val,
};

#[derive(Trace)]
struct EvaluateNamedThunk {
	ctx: Pending<Context>,
	name: IStr,
	value: LocExpr,
}

impl ThunkValue for EvaluateNamedThunk {
	type Output = Val;
	fn get(self: Box<Self>, s: State) -> Result<Val> {
		evaluate_named(s, self.ctx.unwrap(), &self.value, self.name)
	}
}

/// Creates correct [context](Context) for function body evaluation returning error on invalid call.
///
/// ## Parameters
/// * `ctx`: used for passed argument expressions' execution and for body execution (if `body_ctx` is not set)
/// * `body_ctx`: used for default parameter values' execution and for body execution (if set)
/// * `params`: function parameters' definition
/// * `args`: passed function arguments
/// * `tailstrict`: if set to `true` function arguments are eagerly executed, otherwise - lazily
pub fn parse_function_call(
	s: State,
	ctx: Context,
	body_ctx: Context,
	params: &ParamsDesc,
	args: &dyn ArgsLike,
	tailstrict: bool,
) -> Result<Context> {
	let mut passed_args = GcHashMap::with_capacity(params.len());
	if args.unnamed_len() > params.len() {
		throw!(TooManyArgsFunctionHas(params.len()))
	}

	let mut filled_named = 0;
	let mut filled_positionals = 0;

	args.unnamed_iter(s.clone(), ctx.clone(), tailstrict, &mut |id, arg| {
		let name = params[id].0.clone();
		destruct(
			&name,
			arg,
			Pending::new_filled(ctx.clone()),
			&mut passed_args,
		)?;
		filled_positionals += 1;
		Ok(())
	})?;

	args.named_iter(s, ctx, tailstrict, &mut |name, value| {
		// FIXME: O(n) for arg existence check
		if !params.iter().any(|p| p.0.name().as_ref() == Some(name)) {
			throw!(UnknownFunctionParameter((name as &str).to_owned()));
		}
		if passed_args.insert(name.clone(), value).is_some() {
			throw!(BindingParameterASecondTime(name.clone()));
		}
		filled_named += 1;
		Ok(())
	})?;

	if filled_named + filled_positionals < params.len() {
		// Some args are unset, but maybe we have defaults for them
		// Default values should be created in newly created context
		let fctx = Context::new_future();
		let mut defaults =
			GcHashMap::with_capacity(params.len() - filled_named - filled_positionals);

		for (idx, param) in params.iter().enumerate().filter(|p| p.1 .1.is_some()) {
			if let Some(name) = param.0.name() {
				if passed_args.contains_key(&name) {
					continue;
				}
			} else if idx < filled_positionals {
				continue;
			}

			destruct(
				&param.0,
				Thunk::new(tb!(EvaluateNamedThunk {
					ctx: fctx.clone(),
					name: param.0.name().unwrap_or_else(|| "<destruct>".into()),
					value: param.1.clone().expect("default exists"),
				})),
				fctx.clone(),
				&mut defaults,
			)?;
			if param.0.name().is_some() {
				filled_named += 1;
			} else {
				filled_positionals += 1;
			}
		}

		// Some args still weren't filled
		if filled_named + filled_positionals != params.len() {
			for param in params.iter().skip(args.unnamed_len()) {
				let mut found = false;
				args.named_names(&mut |name| {
					if Some(name) == param.0.name().as_ref() {
						found = true;
					}
				});
				if !found {
					throw!(FunctionParameterNotBoundInCall(
						param
							.0
							.clone()
							.name()
							.unwrap_or_else(|| "<destruct>".into())
					));
				}
			}
			unreachable!();
		}

		Ok(body_ctx
			.extend(passed_args, None, None, None)
			.extend(defaults, None, None, None)
			.into_future(fctx))
	} else {
		let body_ctx = body_ctx.extend(passed_args, None, None, None);
		Ok(body_ctx)
	}
}

/// You shouldn't probally use this function, use `jrsonnet_macros::builtin` instead
///
/// ## Parameters
/// * `ctx`: used for passed argument expressions' execution and for body execution (if `body_ctx` is not set)
/// * `params`: function parameters' definition
/// * `args`: passed function arguments
/// * `tailstrict`: if set to `true` function arguments are eagerly executed, otherwise - lazily
pub fn parse_builtin_call(
	s: State,
	ctx: Context,
	params: &[BuiltinParam],
	args: &dyn ArgsLike,
	tailstrict: bool,
) -> Result<GcHashMap<BuiltinParamName, Thunk<Val>>> {
	let mut passed_args = GcHashMap::with_capacity(params.len());
	if args.unnamed_len() > params.len() {
		throw!(TooManyArgsFunctionHas(params.len()))
	}

	let mut filled_args = 0;

	args.unnamed_iter(s.clone(), ctx.clone(), tailstrict, &mut |id, arg| {
		let name = params[id].name.clone();
		passed_args.insert(name, arg);
		filled_args += 1;
		Ok(())
	})?;

	args.named_iter(s, ctx, tailstrict, &mut |name, arg| {
		// FIXME: O(n) for arg existence check
		let p = params
			.iter()
			.find(|p| p.name == name as &str)
			.ok_or_else(|| UnknownFunctionParameter((name as &str).to_owned()))?;
		if passed_args.insert(p.name.clone(), arg).is_some() {
			throw!(BindingParameterASecondTime(name.clone()));
		}
		filled_args += 1;
		Ok(())
	})?;

	if filled_args < params.len() {
		for param in params.iter().filter(|p| p.has_default) {
			if passed_args.contains_key(&param.name) {
				continue;
			}
			filled_args += 1;
		}

		// Some args still wasn't filled
		if filled_args != params.len() {
			for param in params.iter().skip(args.unnamed_len()) {
				let mut found = false;
				args.named_names(&mut |name| {
					if name as &str == &param.name as &str {
						found = true;
					}
				});
				if !found {
					throw!(FunctionParameterNotBoundInCall(param.name.clone().into()));
				}
			}
			unreachable!();
		}
	}
	Ok(passed_args)
}

/// Creates Context, which has all argument default values applied
/// and with unbound values causing error to be returned
pub fn parse_default_function_call(body_ctx: Context, params: &ParamsDesc) -> Result<Context> {
	#[derive(Trace)]
	struct DependsOnUnbound(IStr);
	impl ThunkValue for DependsOnUnbound {
		type Output = Val;
		fn get(self: Box<Self>, _: State) -> Result<Val> {
			Err(FunctionParameterNotBoundInCall(self.0.clone()).into())
		}
	}

	let fctx = Context::new_future();

	let mut bindings = GcHashMap::new();

	for param in params.iter() {
		if let Some(v) = &param.1 {
			destruct(
				&param.0.clone(),
				Thunk::new(tb!(EvaluateNamedThunk {
					ctx: fctx.clone(),
					name: param.0.name().unwrap_or_else(|| "<destruct>".into()),
					value: v.clone(),
				})),
				fctx.clone(),
				&mut bindings,
			)?;
		} else {
			destruct(
				&param.0,
				Thunk::new(tb!(DependsOnUnbound(
					param.0.name().unwrap_or_else(|| "<destruct>".into())
				))),
				fctx.clone(),
				&mut bindings,
			)?;
		}
	}

	Ok(body_ctx
		.extend(bindings, None, None, None)
		.into_future(fctx))
}
