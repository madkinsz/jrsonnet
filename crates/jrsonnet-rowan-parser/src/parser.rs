use std::{cell::Cell, fmt::Display, rc::Rc};

use miette::{LabeledSpan, SourceOffset, SourceSpan};
use rowan::{GreenNode, TextRange, TextSize};

use crate::{
	binary::BinaryOperator,
	event::{Event, Sink},
	lex::{lex, Lexeme},
	marker::{AsRange, CompletedMarker, Marker, Ranger},
	token_set::SyntaxKindSet,
	unary::UnaryOperator,
	SyntaxKind,
	SyntaxKind::*,
	SyntaxNode, T, TS,
};

pub struct Parse {
	pub green_node: GreenNode,
	pub errors: Vec<SyntaxError>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub enum ExpectedSyntax {
	Named(&'static str),
	Unnamed(SyntaxKind),
}
impl Display for ExpectedSyntax {
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		match self {
			ExpectedSyntax::Named(n) => write!(f, "{}", n),
			ExpectedSyntax::Unnamed(u) => write!(f, "{:?}", u),
		}
	}
}

pub struct Parser<'i> {
	lexemes: &'i [Lexeme<'i>],
	pub offset: usize,
	pub events: Vec<Event>,
	pub entered: u32,
	pub hints: Vec<(u32, TextRange, String)>,
	pub last_error_token: usize,
	expected_syntax: Option<ExpectedSyntax>,
	expected_syntax_tracking_state: Rc<Cell<ExpectedSyntaxTrackingState>>,
}

const DEFAULT_RECOVERY_SET: SyntaxKindSet = TS![; ')' ']' '}' local];

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum SyntaxError {
	Unexpected {
		expected: ExpectedSyntax,
		found: SyntaxKind,
		range: TextRange,
	},
	Missing {
		expected: ExpectedSyntax,
		offset: TextSize,
	},
	Custom {
		error: String,
		range: TextRange,
	},
	Hint {
		error: String,
		range: TextRange,
	},
}

impl Into<LabeledSpan> for SyntaxError {
	fn into(self) -> LabeledSpan {
		match self {
			SyntaxError::Unexpected {
				expected,
				found,
				range,
			} => LabeledSpan::new_with_span(
				Some(format!("expected {}, found {:?}", expected, found)),
				SourceSpan::new(
					SourceOffset::from(usize::from(range.start())),
					SourceOffset::from(usize::from(range.end() - range.start())),
				),
			),
			SyntaxError::Missing { expected, offset } => LabeledSpan::new_with_span(
				Some(format!("missing {}", expected)),
				SourceSpan::new(
					SourceOffset::from(usize::from(offset)),
					SourceOffset::from(0),
				),
			),
			SyntaxError::Custom { error, range } | SyntaxError::Hint { error, range } => {
				LabeledSpan::new_with_span(
					Some(format!("{}", error)),
					SourceSpan::new(
						SourceOffset::from(usize::from(range.start())),
						SourceOffset::from(usize::from(range.end() - range.start())),
					),
				)
			}
		}
	}
}

impl<'i> Parser<'i> {
	fn new(lexemes: &'i [Lexeme<'i>]) -> Self {
		Self {
			lexemes,
			offset: 0,
			events: vec![],
			entered: 0,
			last_error_token: 0,
			hints: vec![],
			expected_syntax: None,
			expected_syntax_tracking_state: Rc::new(Cell::new(
				ExpectedSyntaxTrackingState::Unnamed,
			)),
		}
	}
	pub fn clear_outdated_hints(&mut self) {
		let amount = self
			.hints
			.iter()
			.rev()
			.take_while(|h| h.0 > self.entered)
			.count();
		self.hints.truncate(self.hints.len() - amount)
	}
	fn clear_expected_syntaxes(&mut self) {
		self.expected_syntax = None;
		self.expected_syntax_tracking_state
			.set(ExpectedSyntaxTrackingState::Unnamed);
	}
	pub fn start(&mut self) -> Marker {
		let start_event_idx = self.events.len();
		self.events.push(Event::Placeholder);
		self.entered += 1;
		Marker::new(start_event_idx, self.offset)
	}
	pub fn start_ranger(&mut self) -> Ranger {
		let pos = self.offset;
		Ranger { pos }
	}
	fn parse(mut self) -> Vec<Event> {
		let m = self.start();
		expr(&mut self);
		if !self.at_end() {
			let ranger = self.start_ranger();

			while self.peek().is_some() {
				self.bump()
			}
			let end = ranger.finish(&self);
			self.custom_error(end, "unexpected input after expression");
		}
		m.complete(&mut self, SOURCE_FILE);

		self.events
	}

	pub(crate) fn expect(&mut self, kind: SyntaxKind) {
		self.expect_with_recovery_set(kind, TS![])
	}

	pub(crate) fn expect_with_recovery_set(
		&mut self,
		kind: SyntaxKind,
		recovery_set: SyntaxKindSet,
	) {
		if self.at(kind) {
			self.bump();
		} else {
			self.error_with_recovery_set(recovery_set);
		}
	}

	pub(crate) fn expect_with_no_skip(&mut self, kind: SyntaxKind) {
		if self.at(kind) {
			self.bump();
		} else {
			self.error_with_no_skip();
		}
	}
	pub(crate) fn last_token_range(&self) -> Option<TextRange> {
		self.lexemes.last().map(|Lexeme { range, .. }| *range)
	}
	fn current_token(&self) -> Lexeme<'i> {
		self.lexemes[self.offset]
	}
	fn previous_token(&mut self) -> Option<Lexeme<'i>> {
		if self.offset == 0 {
			return None;
		}
		let mut previous_token_idx = self.offset - 1;
		while self
			.lexemes
			.get(previous_token_idx)
			.map_or(false, |l| l.kind.is_trivia())
			&& previous_token_idx != 0
		{
			previous_token_idx -= 1;
		}

		Some(self.lexemes[previous_token_idx])
	}
	pub fn start_of_token(&self, mut idx: usize) -> TextSize {
		while self.lexemes[idx].kind.is_trivia() {
			idx += 1;
		}
		self.lexemes[idx].range.start()
	}
	pub fn end_of_token(&self, mut idx: usize) -> TextSize {
		while self.lexemes[idx].kind.is_trivia() {
			idx -= 1;
		}
		self.lexemes[idx].range.end()
	}
	pub(crate) fn custom_error(&mut self, marker: impl AsRange, error: impl AsRef<str>) {
		self.last_error_token = marker.end_token();
		self.events.push(Event::Error(SyntaxError::Custom {
			error: error.as_ref().to_string(),
			range: marker.as_range(self),
		}));
	}
	pub(crate) fn error_with_recovery_set(
		&mut self,
		recovery_set: SyntaxKindSet,
	) -> Option<CompletedMarker> {
		self.error_with_recovery_set_no_default(recovery_set.union(DEFAULT_RECOVERY_SET))
	}
	pub fn error_with_no_skip(&mut self) -> Option<CompletedMarker> {
		self.error_with_recovery_set_no_default(SyntaxKindSet::ALL)
	}

	pub fn error_with_recovery_set_no_default(
		&mut self,
		recovery_set: SyntaxKindSet,
	) -> Option<CompletedMarker> {
		let expected_syntax = self.expected_syntax.take().unwrap();
		self.expected_syntax_tracking_state
			.set(ExpectedSyntaxTrackingState::Unnamed);

		if self.at_end() || self.at_set(recovery_set) {
			let range = self
				.previous_token()
				.map(|t| t.range)
				.unwrap_or(TextRange::at(TextSize::from(0), TextSize::from(0)));

			self.events.push(Event::Error(SyntaxError::Missing {
				expected: expected_syntax,
				offset: range.end(),
			}));
			return None;
		}

		let current_token = self.current_token();

		self.events.push(Event::Error(SyntaxError::Unexpected {
			expected: expected_syntax.clone(),
			found: current_token.kind,
			range: current_token.range,
		}));
		self.clear_expected_syntaxes();
		self.last_error_token = self.offset;

		let m = self.start();
		self.bump();
		Some(m.complete(self, SyntaxKind::ERROR))
	}

	fn bump(&mut self) {
		self.skip_trivia();
		assert_ne!(self.offset, self.lexemes.len(), "already at end");
		self.events.push(Event::Token);
		self.offset += 1;
		self.clear_expected_syntaxes();
	}
	fn peek(&mut self) -> Option<SyntaxKind> {
		self.skip_trivia();
		self.peek_raw()
	}
	pub fn peek_token(&mut self) -> Option<&Lexeme<'i>> {
		self.skip_trivia();
		self.peek_token_raw()
	}
	fn skip_trivia(&mut self) {
		while self.peek_raw().map(|c| c.is_trivia()).unwrap_or(false) {
			self.offset += 1;
		}
	}
	fn peek_raw(&mut self) -> Option<SyntaxKind> {
		self.lexemes.get(self.offset).map(|l| l.kind)
	}
	fn peek_token_raw(&mut self) -> Option<&Lexeme<'i>> {
		self.lexemes.get(self.offset)
	}
	#[must_use]
	pub(crate) fn expected_syntax_name(&mut self, name: &'static str) -> ExpectedSyntaxGuard {
		self.expected_syntax_tracking_state
			.set(ExpectedSyntaxTrackingState::Named);
		self.expected_syntax = Some(ExpectedSyntax::Named(name));

		ExpectedSyntaxGuard::new(Rc::clone(&self.expected_syntax_tracking_state))
	}
	pub fn at(&mut self, kind: SyntaxKind) -> bool {
		if let ExpectedSyntaxTrackingState::Unnamed = self.expected_syntax_tracking_state.get() {
			self.expected_syntax = Some(ExpectedSyntax::Unnamed(kind));
		}
		self.peek() == Some(kind)
	}
	pub fn at_set(&mut self, set: SyntaxKindSet) -> bool {
		self.peek().map_or(false, |k| set.contains(k))
	}
	pub fn at_end(&mut self) -> bool {
		self.peek().is_none()
	}
}
pub(crate) struct ExpectedSyntaxGuard {
	expected_syntax_tracking_state: Rc<Cell<ExpectedSyntaxTrackingState>>,
}

impl ExpectedSyntaxGuard {
	fn new(expected_syntax_tracking_state: Rc<Cell<ExpectedSyntaxTrackingState>>) -> Self {
		Self {
			expected_syntax_tracking_state,
		}
	}
}

impl Drop for ExpectedSyntaxGuard {
	fn drop(&mut self) {
		self.expected_syntax_tracking_state
			.set(ExpectedSyntaxTrackingState::Unnamed);
	}
}

#[derive(Debug, Clone, Copy)]
enum ExpectedSyntaxTrackingState {
	Named,
	Unnamed,
}
macro_rules! at_match {
	($p:ident {
		$($r:expr => $e:expr,)*
		_ => $else:expr $(,)?
	}) => {{
		$(
			if $p.at($r) {$e} else
		)* {
			$else
		}
	}}
}

fn expr(p: &mut Parser) {
	expr_binding_power(p, 0);
}
fn expr_binding_power(p: &mut Parser, minimum_binding_power: u8) -> Option<CompletedMarker> {
	let mut lhs = lhs(p)?;

	loop {
		let op = at_match!(p {
			T![*] => BinaryOperator::Mul,
			T![/] => BinaryOperator::Div,
			T![%] => BinaryOperator::Mod,
			T![+] => BinaryOperator::Plus,
			T![-] => BinaryOperator::Minus,
			T![<<] => BinaryOperator::ShiftLeft,
			T![>>] => BinaryOperator::ShiftRight,
			T![<] => BinaryOperator::LessThan,
			T![>] => BinaryOperator::GreaterThan,
			T![<=] => BinaryOperator::LessThanOrEqual,
			T![>=] => BinaryOperator::GreaterThanOrEqual,
			T![==] => BinaryOperator::Equal,
			T![!=] => BinaryOperator::NotEqual,
			T![&] => BinaryOperator::BitAnd,
			T![^] => BinaryOperator::BitXor,
			T![|] => BinaryOperator::BitOr,
			T![&&] => BinaryOperator::And,
			T![||] => BinaryOperator::Or,
			T![in] => BinaryOperator::In,
			T!['{'] => BinaryOperator::ObjectApply,
			_ => break,
		});
		let (left_binding_power, right_binding_power) = op.binding_power();
		if left_binding_power < minimum_binding_power {
			break;
		}

		// Object apply is not a real operator, we dont have something to bump
		if op != BinaryOperator::ObjectApply {
			p.bump();
		}

		let m = lhs.precede(p);
		let parsed_rhs = expr_binding_power(p, right_binding_power).is_some();
		lhs = m.complete(
			p,
			if op == BinaryOperator::ObjectApply {
				EXPR_OBJ_EXTEND
			} else {
				EXPR_BINARY
			},
		);

		if !parsed_rhs {
			break;
		}
	}
	Some(lhs)
}
fn compspec(p: &mut Parser) {
	assert!(p.at(T![for]) || p.at(T![if]));
	if p.at(T![for]) {
		let m = p.start();
		p.bump();
		p.expect(IDENT);
		p.expect(T![in]);
		expr(p);
		m.complete(p, FOR_SPEC);
	} else if p.at(T![in]) {
		let m = p.start();
		p.bump();
		expr(p);
		m.complete(p, IF_SPEC);
	} else {
		unreachable!()
	}
}
fn comma(p: &mut Parser) -> bool {
	if p.at(T![,]) {
		p.bump();
		true
	} else {
		false
	}
}
fn comma_with_alternatives(p: &mut Parser, set: SyntaxKindSet) -> bool {
	if p.at(T![,]) {
		p.bump();
		true
	} else if p.at_set(set) {
		p.expect_with_no_skip(T![,]);
		p.bump();
		true
	} else {
		false
	}
}
fn field_name(p: &mut Parser) {
	let _e = p.expected_syntax_name("field name");
	if p.at(T!['[']) {
		p.bump();
		expr(p);
		p.expect(T![']']);
	} else if p.at(IDENT) {
		p.bump()
	} else {
		p.error_with_recovery_set(TS![;]);
	}
}
fn object(p: &mut Parser) -> CompletedMarker {
	assert!(p.at(T!['{']));
	let m = p.start();
	p.bump();

	loop {
		if p.at(T!['}']) {
			p.bump();
			break;
		}
		let m = p.start();
		field_name(p);
		p.expect(T![,]);
		expr(p);
		while p.at(T![for]) || p.at(T![if]) {
			compspec(p)
		}
		m.complete(p, MEMBER);
		if comma_with_alternatives(p, SyntaxKindSet::new(&[T![=]])) {
			continue;
		}
		p.expect(R_BRACE);
		break;
	}

	m.complete(p, OBJ_BODY)
}

fn params(p: &mut Parser) -> CompletedMarker {
	assert!(p.at(T!['(']));
	let m = p.start();
	p.bump();

	loop {
		if p.at(T![')']) {
			p.bump();
			break;
		}
		let m = p.start();
		p.expect(IDENT);
		if p.at(T![=]) {
			p.bump();
			expr(p);
		}
		m.complete(p, PARAM);
		if comma(p) {
			continue;
		}
		p.expect(T![')']);
		break;
	}

	m.complete(p, PARAMS_DESC)
}
fn args(p: &mut Parser) {
	assert!(p.at(T!['(']));
	p.bump();

	let mut error_positional_start = None::<Marker>;
	let mut started_named = Cell::new(false);
	let mut on_positional = |p: &mut Parser, m: Marker| {
		let c = m.complete(p, ARG);
		if started_named.get() && error_positional_start.is_none() {
			error_positional_start = Some(c.precede(p));
		}
	};
	loop {
		if p.at(T![')']) {
			break;
		}

		let m = p.start();
		if p.at(IDENT) {
			p.bump();
			if p.at(T![=]) {
				p.bump();
				expr(p);
				m.complete(p, ARG);
				started_named.set(true);
			} else {
				on_positional(p, m);
			}
		} else {
			expr(p);
			on_positional(p, m);
		}
		if comma(p) {
			continue;
		}
		break;
	}
	if let Some(error_positional_start) = error_positional_start {
		let c = error_positional_start.complete(p, ERROR);
		p.custom_error(c, "positional arguments can't be placed after named")
	}
	p.expect(T![')']);
}

fn array(p: &mut Parser) -> CompletedMarker {
	assert!(p.at(T!['[']));
	// Start the list node
	let m = p.start();
	p.bump(); // '['

	// This vec will have at most one element in case of correct input
	let mut compspecs = Vec::with_capacity(1);
	let mut elems = 0;

	loop {
		if p.at(T![']']) {
			p.bump();
			break;
		}
		elems += 1;
		expr(p);
		let c = p.start_ranger();
		let mut had_spec = false;
		while p.at(T![for]) || p.at(T![if]) {
			had_spec = true;
			compspec(p)
		}
		if had_spec {
			compspecs.push(c.finish(p));
		}
		if comma(p) {
			continue;
		}
		p.expect(T![']']);
		break;
	}

	if elems > 1 && !compspecs.is_empty() {
		for spec in compspecs {
			p.custom_error(
				spec,
				"compspec may only be used if there is only one array element",
			)
		}

		m.complete(p, EXPR_ARRAY)
	} else if !compspecs.is_empty() {
		m.complete(p, EXPR_ARRAY_COMP)
	} else {
		m.complete(p, EXPR_ARRAY)
	}
}

fn lhs(p: &mut Parser) -> Option<CompletedMarker> {
	let mut lhs = lhs_basic(p)?;

	loop {
		if p.at(T![.]) {
			let m = lhs.precede(p);
			p.bump();
			p.expect(IDENT);
			lhs = m.complete(p, EXPR_INDEX);
		} else if p.at(T!['[']) {
			let m = lhs.precede(p);
			p.bump();
			// Start
			if !p.at(T![:]) {
				expr(p);
			}
			if p.at(T![:]) {
				p.bump();
				// End
				if !p.at(T![']']) && !p.at(T![:]) {
					expr(p);
				}
				if p.at(T![:]) {
					p.bump();
					// Step
					if !p.at(T![']']) {
						expr(p);
					}
				}
			}
			p.expect(T![']']);
			lhs = m.complete(p, EXPR_SLICE);
		} else if p.at(T!['(']) {
			let m = lhs.precede(p);
			args(p);
			lhs = m.complete(p, EXPR_APPLY);
		} else {
			break;
		}
	}

	Some(lhs)
}

fn lhs_basic(p: &mut Parser) -> Option<CompletedMarker> {
	let _e = p.expected_syntax_name("value");
	Some(if p.peek().map(|l| l.is_literal()).unwrap_or(false) {
		let m = p.start();
		p.bump();
		m.complete(p, EXPR_LITERAL)
	} else if p.peek().map(|l| l.is_string()).unwrap_or(false) {
		let m = p.start();
		p.bump();
		m.complete(p, EXPR_STRING)
	} else if p.peek().map(|l| l.is_number()).unwrap_or(false) {
		let m = p.start();
		p.bump();
		m.complete(p, EXPR_NUMBER)
	} else if p.at(IDENT) {
		let m = p.start();
		p.bump();
		m.complete(p, EXPR_VAR)
	} else if p.at(T!['[']) {
		array(p)
	} else if p.at(T!['{']) {
		object(p)
	} else if p.at(T![local]) {
		let m = p.start();
		p.bump();
		let mut sus_local = None;
		loop {
			p.expect_with_recovery_set(IDENT, TS![= ; local]);
			if p.at(T!['(']) {
				params(p);
			}

			let sus_local_candidate = p.start_ranger();
			p.expect_with_recovery_set(T![=], TS![; local]);

			sus_local = p.at(T![local]).then(|| sus_local_candidate.finish(p));
			expr(p);

			if !comma(p) {
				break;
			}
		}
		p.expect(T![;]);
		if let Some(sus_local) = sus_local {
			if sus_local.had_error_since(p) {
				p.custom_error(sus_local, "unusal local placement, missing ';' ?")
			}
		}
		expr(p);
		m.complete(p, T![local])
	} else if p.at(T![function]) {
		let m = p.start();
		p.bump();
		args(p);
		expr(p);
		m.complete(p, EXPR_FUNCTION)
	} else if p.at(T![error]) {
		let m = p.start();
		p.bump();
		expr(p);
		m.complete(p, EXPR_ERROR)
	} else if p.at(T![assert]) {
		let m = p.start();
		p.bump();
		expr(p);
		if p.at(T![:]) {
			p.bump();
			expr(p);
		}
		m.complete(p, EXPR_ASSERT)
	} else if p.at(T![import]) || p.at(T![importstr]) || p.at(T![importbin]) {
		let m = p.start();
		p.bump();
		expr(p);
		m.complete(p, EXPR_IMPORT)
	} else if p.at(T![-]) || p.at(T![!]) || p.at(T![~]) {
		let op = match p.peek().unwrap() {
			T![-] => UnaryOperator::Minus,
			T![!] => UnaryOperator::Not,
			T![~] => UnaryOperator::BitNegate,
			_ => unreachable!(),
		};
		let ((), right_binding_power) = op.binding_power();

		let m = p.start();
		p.bump();
		expr_binding_power(p, right_binding_power);
		m.complete(p, EXPR_UNARY)
	} else if p.at(T!['(']) {
		let m = p.start();
		p.bump();
		expr(p);
		assert!(p.at(T![')']));
		p.bump();
		m.complete(p, EXPR_PARENED)
	} else {
		p.error_with_no_skip();
		return None;
	})
}

impl Parse {
	pub fn syntax(&self) -> SyntaxNode {
		SyntaxNode::new_root(self.green_node.clone())
	}
}

pub fn parse(input: &str) -> Parse {
	let lexemes = lex(input);
	let parser = Parser::new(&lexemes);
	let events = parser.parse();
	dbg!(&events);
	let sink = Sink::new(events, &lexemes);

	sink.finish()
}
