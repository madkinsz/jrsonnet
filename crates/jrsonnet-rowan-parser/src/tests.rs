#![cfg(test)]

use miette::{Diagnostic, GraphicalReportHandler, LabeledSpan};
use thiserror::Error;

use crate::parser::parse;

#[derive(Debug, Error)]
#[error("syntax error")]
struct MyDiagnostic {
	code: String,
	spans: Vec<LabeledSpan>,
}
impl Diagnostic for MyDiagnostic {
	fn code<'a>(&'a self) -> Option<Box<dyn std::fmt::Display + 'a>> {
		None
	}

	fn severity(&self) -> Option<miette::Severity> {
		None
	}

	fn help<'a>(&'a self) -> Option<Box<dyn std::fmt::Display + 'a>> {
		None
	}

	fn url<'a>(&'a self) -> Option<Box<dyn std::fmt::Display + 'a>> {
		None
	}

	fn source_code(&self) -> Option<&dyn miette::SourceCode> {
		Some(&self.code)
	}

	fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
		Some(Box::new(self.spans.clone().into_iter()))
	}

	fn related<'a>(&'a self) -> Option<Box<dyn Iterator<Item = &'a dyn Diagnostic> + 'a>> {
		None
	}
}

fn process(text: &str) -> String {
	use std::fmt::Write;
	let mut out = String::new();
	let node = parse(text);
	write!(out, "{:#?}", node.syntax()).unwrap();
	if !node.errors.is_empty() && !text.is_empty() {
		writeln!(out, "===").unwrap();
		for err in &node.errors {
			writeln!(out, "{:?}", err).unwrap();
		}
		let diag = MyDiagnostic {
			code: text.to_string(),
			spans: node.errors.into_iter().map(|e| e.into()).collect(),
		};

		let handler = GraphicalReportHandler::new();

		write!(out, "===").unwrap();
		handler.render_report(&mut out, &diag).unwrap();
	}
	out
}
macro_rules! mk_test {
		($($name:ident => $test:expr)+) => {$(
			#[test]
			fn $name() {
				let src = indoc::indoc!($test);
				let result = process(&src);
				insta::assert_snapshot!(stringify!($name), result, src);

			}
		)+};
	}
mk_test!(
	empty => r#" "#
	function => r#"
			function(a, b = 1) a + b
		"#
	function_error_no_value => r#"
			function(a, b = ) a + b
		"#
	function_error_rparen => r#"
			function(a, b
		"#
	function_error_body => r#"
			function(a, b)
		"#
	local_novalue => r#"
			local a =
		"#
	local_no_value_recovery => r#"
			local a =
			local b = 3;
			1
		"#

	array_comp => r#"
			[a for a in [1, 2, 3]]
		"#
	array_comp_incompatible_with_multiple_elems => r#"
			[a for a in [1, 2, 3], b]
		"#

	no_rhs => r#"
			a +
		"#
	no_lhs => r#"
			+ 2
		"#
	no_operator => "
			2 2
		"

	named_before_positional => "
			a(1, 2, b=4, 3, 5, k = 12, 6)
		"

	wrong_field_end => "
			{
				a: 1;
				b: 2;
			}
		"
);
