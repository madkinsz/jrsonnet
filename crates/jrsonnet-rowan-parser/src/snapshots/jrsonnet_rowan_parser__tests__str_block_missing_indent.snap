---
source: crates/jrsonnet-rowan-parser/src/tests.rs
expression: "|||\n"
---
SOURCE_FILE@0..4
  EXPR_STRING@0..4
    ERROR_STRING_BLOCK_MISSING_INDENT@0..4 "|||\n"

